Python/ceval.c

```diff
@@ -4358,13 +4351,93 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
                 }
                 _PyErr_Clear(tstate);
             }
+        iterator_exhausted_no_error:
             /* iterator ended normally */
-            STACK_SHRINK(1);
-            Py_DECREF(iter);
-            JUMPBY(oparg);
+            assert(!_PyErr_Occurred(tstate));
+            Py_DECREF(POP());
+            JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER + oparg);
             DISPATCH();
         }
 
+        TARGET(FOR_ITER_ADAPTIVE) {
+            assert(cframe.use_tracing == 0);
+            _PyForIterCache *cache = (_PyForIterCache *)next_instr;
+            if (cache->counter == 0) {
+                next_instr--;
+                _Py_Specialize_ForIter(TOP(), next_instr);
+                NOTRACE_DISPATCH_SAME_OPARG();
+            }
+            else {
+                STAT_INC(FOR_ITER, deferred);
+                cache->counter--;
+                JUMP_TO_INSTRUCTION(FOR_ITER);
+            }
+        }
+
+        TARGET(FOR_ITER_LIST) {
+            assert(cframe.use_tracing == 0);
+            _PyListIterObject *it = (_PyListIterObject *)TOP();
+            DEOPT_IF(Py_TYPE(it) != &PyListIter_Type, FOR_ITER);
+            STAT_INC(FOR_ITER, hit);
+            PyListObject *seq = it->it_seq;
+            if (seq == NULL) {
+                goto iterator_exhausted_no_error;
+            }
+            if (it->it_index < PyList_GET_SIZE(seq)) {
+                PyObject *next = PyList_GET_ITEM(seq, it->it_index++);
+                Py_INCREF(next);
+                PUSH(next);
+                JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER);
+                NOTRACE_DISPATCH();
+            }
+            it->it_seq = NULL;
+            Py_DECREF(seq);
+            goto iterator_exhausted_no_error;
+        }
+
+        TARGET(FOR_ITER_RANGE) {
+            assert(cframe.use_tracing == 0);
+            _PyRangeIterObject *r = (_PyRangeIterObject *)TOP();
+            DEOPT_IF(Py_TYPE(r) != &PyRangeIter_Type, FOR_ITER);
+            STAT_INC(FOR_ITER, hit);
+            _Py_CODEUNIT next = next_instr[INLINE_CACHE_ENTRIES_FOR_ITER];
+            assert(_PyOpcode_Deopt[_Py_OPCODE(next)] == STORE_FAST);
+            PyObject **local_ptr = &GETLOCAL(_Py_OPARG(next));
+            PyObject *local = *local_ptr;
+            if (r->index >= r->len) {
+                goto iterator_exhausted_no_error;
+            }
+            JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER + 1);
+            sdigit value = r->start + (digit)(r->index++) * r->step;
+            if (value < _PY_NSMALLPOSINTS && value >= -_PY_NSMALLNEGINTS) {
+                *local_ptr = Py_NewRef(&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS+value]);
+                Py_XDECREF(local);
+                NOTRACE_DISPATCH();
+            }
```

Comment:
Can you please add a comment here explaining this special case? Something like
> This is a special case where a value from a range is stored in a variable that isn't referenced from elsewhere.
> So we can re-use the pre-existing long object and avoid an allocation.

Also just so that I'm understanding things, you don't need to check whether the value can fit in the existing long because it's `sdigit` right?

Reply:
> Also just so that I'm understanding things, you don't need to check whether the value can fit in the existing long because it's sdigit right?

Unfortunately, having the right c type isn't enough, the value has to live in the range of possible values `-PyLong_MASK <= x <= PyLong_MASK` for a one-digit PyLongObject. To make computations work, a C `sdigit` has enough room for a sum or difference of logical digits. On 64-bit machines, an `sdigit` is 32 bits, but longs are stored in base-`(2**30)`. So this PR limits `_PyRangeIterObject`s to only have values `-2**30 < x < 2**30`.